stdlib/fun.ml

```diff
@@ -17,3 +17,18 @@ external id : 'a -> 'a = "%identity"
 let const c _ = c
 let flip f x y = f y x
 let negate p v = not (p v)
+
+exception Finally_raised of exn
+
+let protect ~(finally : unit -> unit) work =
+  let finally_no_exn () =
+    try finally () with e ->
+      let bt = Printexc.get_raw_backtrace () in
+      Printexc.raise_with_backtrace (Finally_raised e) bt
```

Comment:
I had expected to actually have `exception Finally_raised of exn * backtrace` ? What kind of rendering will that give us if we print the backtrace ? Will it not make it as if `Finally_raised` was raised by `finally` ?

Reply:
Yes, that was the reason for my hesitation. With your approach, will you inspect recursively the Finally_raised exception to get to the most precise backtrace (or print them all), e.g. if you catch:
```ocaml
Finally_raised (Finally_raised (e, bt1), bt2)
```
?

Reply:
> Yes

I think we should err on the side of being as less confusing as possible and would thus prefer to have what I suggest. I think it's clearer: you get the exception with the backtrace up to the finally and then the Finally exception with the backtrace starting from the finally. And Yes I would print them all.

Reply:
This makes sense; I do not have any strong opinion on this issue. While I agree with your points, it still looks somewhat clumsy to me to give `Finally_raised` a complicated signature just to get a list of backtraces which are all prefixes of each other; surely there is a better way. I would like to know what other developers prefer and/or if they have other ideas.

Reply:
To develop a bit the idea behind my implementation, you only really care about the longest backtrace. If I am not mistaken, the information given by the various prefixes is redundant with the information given by the various occurrences of `Re-raised at file "fun.ml", line 27, characters 36-54` in the backtrace.

Now you say that you are ready to inspect the `Finally_raised` exception for display purposes, but then your concern about my implementation could also be addressed by having `Finally_raised (…(Finally_raised e)…)` be printed in a way that reflects its special meaning, e.g. `Exception: <e> under <n> Finally_raised`. (The number is not informative, we could also make sure that the exception is not wrapped if it is already a `Finally_raised`.)

Reply:
Shouldn't the `Finally_raised` value record both exceptions and at least one backtrace? Both exceptions and both backtraces are useful because you might want to debug both your `finally` and your `work` functions.